File: leet/20e_ValidParantheses.java
import java.util.*;

public class Solution {
    public boolean isValid(String s) {
        if(s==null) return false;
        
        Stack<Character> st = new Stack<>();
        int len = s.length();
        
        if(len%2==1)    return false;
        
        for(int i=0; i<len; i++) {
            char c = s.charAt(i);
            
            // open brace, push to Stack
            if(c=='(' || c=='{' || c=='[') {
                st.push(c);
            } else {
                if(st.size()==0)    return false;
                
                char d = st.pop();
                if(c==')' && d!='(')    return false;
                if(c=='}' && d!='{')    return false;
                if(c==']' && d!='[')    return false;
            }
        }
        
        if(st.size()>0) return false;
        
        return true;
    }
}

File: leet/20e_Valid_Parantheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for c in s:
            if c in ('(', '{', '['):
                stack.append(c)
            else:
                if not stack:
                    return False
                
                left = stack.pop()
                if not self.match(left, c):
                    return False
        
        if stack:
            return False
        
        return True
    
    
    def match(self, a, b):
        if a == '(':
            return b == ')'
        if a == '{':
            return b == '}'
        if a == '[':
            return b == ']'

if __name__ == "__main__":
    s = Solution()
    print(s.isValid("()"))
    print(s.isValid("()[]{}"))
    print(s.isValid("(]"))
    print(s.isValid("([)]"))
    print(s.isValid("{[]}"))
    print(s.isValid("["))

